### PR TITLE
FMWK-86 Fix adding a new list element

### PR DESCRIPTION
**Pull request by agrgr**
Fix for adding a new list element correctly within post-production (in case of 2-step operations).

### CODE REVIEW

File: src/main/java/com/aerospike/documentapi/pathparts/MapPathPart.java [hunk 1]
```diff
@@ -52,7 +52,10 @@ public Operation toAerospikeGetOperation(String binName, CTX[] contexts) {
     }
 
     @Override
-    public Operation toAerospikePutOperation(String binName, Object object, CTX[] contexts) {
+    public Operation toAerospikePutOperation(String binName, Object object, CTX[] contexts) throws IllegalArgumentException {
+        if (object.getClass().isArray()) {
+            throw new IllegalArgumentException("Unable to put an array, expecting a list or a map");
```
Comment (reugn) [inline src/main/java/com/aerospike/documentapi/pathparts/MapPathPart.java:57]: It also can be a regular object.

File: src/main/java/com/aerospike/documentapi/pathparts/ListPathPart.java [hunk 2]
```diff
@@ -51,7 +51,10 @@ public Operation toAerospikeGetOperation(String binName, CTX[] contexts) {
     }
 
     @Override
-    public Operation toAerospikePutOperation(String binName, Object object, CTX[] contexts) {
+    public Operation toAerospikePutOperation(String binName, Object object, CTX[] contexts) throws IllegalArgumentException {
+        if (object.getClass().isArray()) {
+            throw new IllegalArgumentException("The given object is not of type " + getType());
```
Comment (reugn) [inline src/main/java/com/aerospike/documentapi/pathparts/ListPathPart.java:56]: Same as at `MapPathPart`?

File: src/main/java/com/aerospike/documentapi/JsonPathQuery.java [hunk 3]
```diff
@@ -22,21 +23,38 @@ public static Object putOrSet(JsonPathObject jsonPathObject, Object object, Obje
...
         String jsonPath = DOCUMENT_ROOT + jsonPathObject.getJsonPathSecondStepQuery();
         JSONArray keys = JsonPath.parse(resultJson).read(jsonPath);
-        if (!keys.isEmpty()) {
-            return JsonPath.parse(resultJson).set(jsonPath, value).json();
+        // if jsonPath exists or if it leads to an array element
+        if (!keys.isEmpty() || jsonPath.charAt(jsonPath.length() - 1) == ']') {
+            return setOrAdd(resultJson, jsonPath, value);
         }
+        // if jsonPath does not exist in json, and it leads to a map element
+            return put(resultJson, jsonPath, value);
```
Comment (reugn) [inline src/main/java/com/aerospike/documentapi/JsonPathQuery.java:31]: Code format.

File: src/test/java/com/aerospike/documentapi/DocumentAPITest.java [hunk 4]
```diff
@@ -253,6 +253,15 @@ public void testPositivePut() throws IOException,
...
      */
+
+    /**
+     * Check correct response to erroneous access
+     * <p>
+     * Putting a key into a map that doesn't exist
+     * Putting a value into a list that doesn't exist
+     * Treating a map as if it were a list
+     * Treating a list as if it were a map
+     */
```
Comment (reugn) [inline src/test/java/com/aerospike/documentapi/DocumentAPITest.java:264]: Javadoc duplication.

File: src/main/java/com/aerospike/documentapi/pathparts/PathPart.java [hunk 5]
```diff
@@ -13,9 +13,12 @@ public abstract class PathPart {
 
     public abstract Operation toAerospikeGetOperation(String binName, CTX[] contexts);
 
-    public abstract Operation toAerospikePutOperation(String binName, Object object, CTX[] contexts);
+    public abstract Operation toAerospikePutOperation(String binName, Object object, CTX[] contexts) throws IllegalArgumentException;
 
-    public Operation toAerospikeAppendOperation(String binName, Object object, CTX[] contexts) {
+    public Operation toAerospikeAppendOperation(String binName, Object object, CTX[] contexts) throws IllegalArgumentException {
+        if (object.getClass().isArray()) {
+            throw new IllegalArgumentException("Unable to append an array, expecting a list or a map");
```
Comment (reugn) [inline src/main/java/com/aerospike/documentapi/pathparts/PathPart.java:20]: Same as at `MapPathPart`?

File: src/test/java/com/aerospike/documentapi/DocumentAPITest.java (hunk 4 above)
Comment (agrgr) [reply src/test/java/com/aerospike/documentapi/DocumentAPITest.java:264]: Merge artifact, fixed.

File: src/main/java/com/aerospike/documentapi/JsonPathQuery.java (hunk 3 above)
Comment (agrgr) [reply src/main/java/com/aerospike/documentapi/JsonPathQuery.java:31]: Done

File: src/main/java/com/aerospike/documentapi/pathparts/MapPathPart.java (hunk 1 above)
Comment (agrgr) [reply src/main/java/com/aerospike/documentapi/pathparts/MapPathPart.java:57]: Yes, what I was trying to say is if a user tries to pass an array then he/she is likely to use a list or a map instead.

File: src/main/java/com/aerospike/documentapi/pathparts/PathPart.java (hunk 5 above)
Comment (agrgr) [reply src/main/java/com/aerospike/documentapi/pathparts/PathPart.java]: Extracted to a separate method.

File: src/main/java/com/aerospike/documentapi/pathparts/ListPathPart.java (hunk 2 above)
Comment (agrgr) [reply src/main/java/com/aerospike/documentapi/pathparts/ListPathPart.java]: Extracted to a separate method.

File: src/main/java/com/aerospike/documentapi/pathparts/MapPathPart.java (hunk 1 above)
Comment (agrgr) [reply src/main/java/com/aerospike/documentapi/pathparts/MapPathPart.java]: Rephrased.

**Review by reugn — APPROVED** [review]

**Pull request merged**